src/identity-wallet.ts

```diff
@@ -0,0 +1,526 @@
+  }
+
+  /**
+   * Get the DIDProvider
+   *
+   * @return    {DidProvider}                   The DIDProvider for this IdentityWallet instance
+   */
+  getDidProvider() {
+    return new DidProvider(this)
+  }
+
+  /**
+   * Determine if consent has been given for spaces for a given origin
+   *
+   * @param     {Array<String>}     spaces          The desired spaces
+   * @param     {String}            origin          Application domain
+   * @param     {String}            opt.address     Optional address (managementKey) if keyring not available yet
+   * @return    {Boolean}                           True if consent has already been given
+   */
+  hasConsent(
+    spaces: Array<string> = [],
+    origin?: string | null,
+    { address }: { address?: string } = {}
+  ): boolean {
+    const key = address ?? this._keyring!.getPublicKeys().managementKey ?? ''
+    const prefix = `3id_consent_${key}_${origin ?? ''}_`
+    const consentExists = (space = '') => Boolean(store.get(prefix + space))
+    return spaces.reduce(
+      (acc, space) => acc && consentExists(space),
+      consentExists()
+    )
+  }
+
+  /**
+   *  Get consent for given spaces for a given origin
+   *
+   * @param     {Array<String>}     spaces          The desired spaces
+   * @param     {String}            origin          Application domain
+   * @param     {String}            opt.address     Optional address (managementKey) if keyring not available yet
+   * @return    {Boolean}                           True consent was given
+   */
+  async getConsent(
+    spaces: Array<string> = [],
+    origin?: string | null,
+    { address }: { address?: string } = {}
+  ) {
+    if (!this.hasConsent(spaces, origin, { address })) {
+      const consent = await this._getConsent({
+        type: 'authenticate',
+        origin,
+        spaces,
+        opts: {
+          address,
+        },
+      })
+      if (!consent) return false
+      const key = address ?? this._keyring!.getPublicKeys().managementKey ?? ''
+      const prefix = `3id_consent_${key}_${origin ?? ''}_`
+      const saveConsent = (space = '') => store.set(prefix + space, true)
+      saveConsent()
+      spaces.map(saveConsent)
+    }
+    return true
+  }
+
+  async getLink(): Promise<string> {
+    if (this._seed) {
+      this._keyring = new Keyring(this._seed)
+      this.DID = await this._get3id()
+      delete this._seed
+      await this._linkManagementAddress()
+    }
+    // for external auth keyring will already exist at this point
+    return this._keyring
+      ? (this._keyring.getPublicKeys().managementKey as string)
+      : Keyring.walletForAuthSecret(this._authSecret!).address
+  }
+
+  getRootSigner(pubKeyId?: string): Signer {
+    if (pubKeyId == null) {
+      return this._keyring!.getRootSigner()
+    }
+
+    const [did, keyId] = pubKeyId.split('#')
+    if (this.DID == null || did !== this.DID) {
+      throw new Error('Invalid DID')
+    }
+    return this._keyring!.getRootSigner(keyId)
```

Comment:
What does the `!` syntax do?

Reply:
It tells the TypeScript compiler to ignore the fact the object is possibly null or undefined, as it's defined as such here: https://github.com/3box/identity-wallet-js/pull/56/files#diff-543d3968cda554f944babf9724a7a039R34
Ideally we shouldn't use it because there should be proper guards for this, but I wanted to limit the scope of the refactoring to avoid making it even larger, so I've used this and `@ts-ignore` comments to avoid compilation errors.

Reply:
Makes sense. Thanks for clarifying!